feat(git): improve readability and maintainability

PFiS1737 opened the pull request:
### Description

- Improve readability and maintainability.
- Default usage of `nf-oct-diff` series icons from Nerdfonts and terminal colors (instead of `RGB` colors).

sxyazi commented on git.yazi/README.md:
```diff
@@ -37,40 +79,47 @@ run  = "git"
-THEME.git = THEME.git or {}
-THEME.git.modified = ui.Style():fg("blue")
-THEME.git.deleted = ui.Style():fg("red"):bold()
+th.git = th.git or {}
```
Please use the original `THEME` here — it's intentional because `th` is read-only, and you can't change its value

PFiS1737 replied:
I'm not sure. I intended to keep this example because I initially thought they were the same. However, when I tested it, I found that it only worked if I used the same variable in both `init.lua` and the plugin (either `THEME` or `th` at the same time). If different variables were used — such as `THEME` in `init.lua` and `th` in the plugin's `main.lua` (or vice versa) — the configuration wouldn't take effect.

I'm not sure if this is my issue (version 25.2.26). I considered using `THEME` everywhere, but since you mentioned that it was marked as deprecated, I used `th` directly.

sxyazi replied:
`THEME` and `th` are not the same, changing the value of `th` doesn't apply to `THEME`.

If we want to completely switch to `th`, it must be introduced as a breaking change because users might already be using `THEME` in their configuration.

sxyazi commented on git.yazi/README.md:
```diff
@@ -15,12 +16,53 @@ ya pack -a yazi-rs/plugins:git
+  linemode = true,
+  linemode_order = 1500,
+  highlight = false,
+  styles = {
```
Using `THEME` to configure the theme instead of the `setup` function is intentional, as I plan to support custom styles in `theme.toml`, see https://github.com/yazi-rs/plugins/issues/44 - users will be able to configure `[git]` in their `theme.toml`, and introducing `setup` would cause ambiguity.

PFiS1737 replied:
I made this decision after reading #44, as some people might prefer to write the configuration of the same plugin together and manage it through a single setup, without having to search across multiple files. I have mentioned this in the README. If you still request a change, I will comply.

sxyazi commented:
This PR contains too much content, which makes it difficult to diff. Is it possible to split it into several smaller PRs?

PFiS1737 commented:
Sorry, this might be quite difficult as it involves an almost complete refactor. However, I can provide some technical details (which I will add in this comment).

PFiS1737 committed:
feat: improve the retrieval of the ignored status

PFiS1737 committed:
feat: improve default config

PFiS1737 committed:
feat: improve readability

PFiS1737 commented:
If you don't mind, I will force push to the PR after I finish. The new commits include the following changes:

- Reverted the configuration options for `icons` and `styles` in `setup` (but I have kept the related commits locally and am ready to patch them anytime — trust me, people tend to prefer keeping configurations together).
- Used `THEME` consistently in both the `README` and `main.lua`, as I still couldn't get different variables to work together (`THEME` in `init.lua` and `th` in `main.lua`) — as I mentioned above. (with the latest commit [38d5b2f](https://github.com/sxyazi/yazi/commit/38d5b2fc877e3421ddc6bc9822a932abaf81e874))
- Unified variable names.
- Added caching in `get_status` (previously `get_kind`) to improve performance.

sxyazi reviewed (changes requested):
[This commit](https://github.com/yazi-rs/plugins/pull/78/commits/ce205d29b540c01ebc3cdbc045eea07c9bb1e5a1)'s changes don't make sense. 

It moves work that was originally handled in async tasks into a sync context, and for each linemode miss, it repeatedly loops to check its parent. Since linemode is invoked for all files on the page, this will lead to a significant amount of memory allocation. 

The bottom line here is that any computational work should be avoided as much as possible during UI rendering.

PFiS1737 commented:
Sorry, I didn't realize that. I will try to revert it. (Perhaps there's a better way to handle this? The introduction of 30 has increased the complexity and difficulty of understanding the code.)

sxyazi commented:
IMO this PR should just focus on adding the new feature and keep the
changes as small as possible so it’s easier to review. Improving code
readability is great, but it might be worth doing that in a separate PR
where we can dig into it a bit more and discuss it further.


On Mon, Mar 3, 2025 at 3:04 PM PFiS ***@***.***> wrote:

> Sorry, I didn't realize that. I will try to revert it. (Perhaps there's a
> better way to handle this? The introduction of 30 has increased the
> complexity and difficulty of understanding the code.)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/yazi-rs/plugins/pull/78#issuecomment-2693457450>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEFWFIHLZSUJZ5BOL6EAUCL2SP5GZAVCNFSM6AAAAABYE6ER3WVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJTGQ2TONBVGA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
> [image: PFiS1737]*PFiS1737* left a comment (yazi-rs/plugins#78)
> <https://github.com/yazi-rs/plugins/pull/78#issuecomment-2693457450>
>
> Sorry, I didn't realize that. I will try to revert it. (Perhaps there's a
> better way to handle this? The introduction of 30 has increased the
> complexity and difficulty of understanding the code.)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/yazi-rs/plugins/pull/78#issuecomment-2693457450>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEFWFIHLZSUJZ5BOL6EAUCL2SP5GZAVCNFSM6AAAAABYE6ER3WVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJTGQ2TONBVGA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

PFiS1737 committed:
feat: revert

PFiS1737 commented:
The rollback is complete, and I have translated the original logic according to the latest type definitions. I will continue to explore how to remove the `__ignored` and `""` entries in `dirs_to_repo`, which are confusing. Or I will try adding comments.

PFiS1737 committed:
feat: update

PFiS1737 commented:
Well, I failed — I couldn't come up with any method other than using an additional flag. 

However, I made some optimizations:

- remove the conditional checks in `bubble_up` and `propagate_down` since they always pass
- add comments and use named flags to make the code a bit easier to understand
- and some minor adjustments

sxyazi commented on git.yazi/main.lua:
```diff
@@ -48,119 +107,132 @@ local function root(cwd)
+	for _, path in ipairs(ignored) do
+		if rel:starts_with(path) then
+			-- if `cwd` is a subfolder of an ignored directory, mark the `cwd` as `__ignored_dir__`
+			new[tostring(cwd)] = "__ignored_dir__"
```
string is not an immediate value, which means that transferring them between threads requires constant memory reallocation, and it also actually occupies more memory than just a number.

sxyazi commented on git.yazi/main.lua:
```diff
@@ -1,43 +1,102 @@
+}
+
+---@type Config
+local Config = {
```
What does capitalizing the first letter mean? Is it a constant or a variable? It seems that the entire repository does not follow such a naming convention

PFiS1737 replied:
Oh, that's a habit of mine — I usually treat config as an independent module (when writing Neovim plugins), and capitalizing the first letter helps emphasize its independence. Perhaps we should change this naming to make it consistent with the others?

sxyazi commented on git.yazi/main.lua:
```diff
@@ -1,43 +1,102 @@
+}
+
+---@type table<string, Status>
+local PATTERNS = {
```
Any reason to change their order?

PFiS1737 replied:
This decision was made based on importance: `untracked` should have a higher priority to serve as a prompt, and `added` should be more important than `modified`. Although `ignored` is handled separately, I assigned it the highest priority to align with the context's sorting method. I also referred to the priority used in NeoTree (a Neovim plugin).

sxyazi commented on git.yazi/main.lua:
```diff
@@ -169,39 +241,43 @@ local function fetch(_, job)
-	-- stylua: ignore
 	local output, err = Command("git")
-		:cwd(tostring(cwd))
-		:args({ "--no-optional-locks", "-c", "core.quotePath=", "status", "--porcelain", "-unormal", "--no-renames", "--ignored=matching" })
```
Here `stylua: ignore` was intentional, please keep the original style.

sxyazi commented on git.yazi/main.lua:
```diff
@@ -169,39 +241,43 @@ local function fetch(_, job)
-		else
-			changed[path] = sign
+		local status, path = match(line)
+		if status and path then
```
Why is this `if` statement required?

PFiS1737 replied:
According to the logic, `match` will return `nil`, and the LSP also suggests performing a nil check.

PFiS1737 committed:
make sure to use tab instead of space

sxyazi commented:
I have only reviewed part of it, and it is still very difficult to review as the different changes are intertwined and this PR refactored almost the entire plugin, and I am not sure how to correctly review them.

I'll try to split these changes into several smaller PRs later and request you review them. I will add you as a collaborator to make sure you get the credits.

PFiS1737 commented:
Sincerely appreciate your efforts and thank you for your patience these days. I will do my best to fulfill my part.

sxyazi commented on git.yazi/main.lua:
```diff
@@ -48,119 +108,134 @@ local function root(cwd)
-
-	-- Chosen by ChatGPT fairly, PRs are welcome to adjust them
-	local t = THEME.git or {}
-	local styles = {
```
`styles` and `signs` are local variables rather than being extracted to a global `Config` intentionally because they are only used in setup (sync plugin). 

This avoids initializing a bunch of unnecessary `ui.Style`s when scheduled as an async fetcher, where they won't be used at all.

PFiS1737 replied:
I will make the changes as requested.

Additionally, why don't we prefer setting `styles` in the default configuration and `README` as `table` instead of `ui.Style`? This would simplify the process for users to move configurations into `theme.toml` (just copy-pasting instead of translating) and would also avoid unnecessary `ui.Style`s.

sxyazi replied:
If you want to represent `ui.Style():bold()` as a Lua table, it would look like `{ modifier = 1 }`, where `1` is the binary value of the `bold` style, which is not user-friendly.

PFiS1737 commented:
Hi, I'm planning to submit separate PRs over the next few days to complete the updates:

- Simplify the code — mainly by removing unnecessary logic, possibly with some variable name changes.
- Add type declarations and comments.
- Adjust the priority of codes (`untracked` higher, perhaps needs discussion?).
- Modify the default `styles` and `signs` to use terminal colors and consistent series icons.
- Add support for `renamed` status.
- Add highlight support for list entries.

Today's will be the first.

You can review them and give feedback whenever you have time, and I'll submit the next one once the previous one is merged.
